### PR TITLE
Queue requests from createRemoteFileNode and control concurrency of requests

### PR DESCRIPTION
**Pull request by tsimons**
The blog I work on has over 9,000 media objects and currently it tries
to download them all simultaneously. This PR chunks them in to groups of 100 by default. The `chunkSize` setting can be used to increase this limit

### CODE REVIEW

**Comment by KyleAMathews** [comment]
How about instead we add a queue to gatsby-source-filesystem? There's a number of plugins that are using createRemoteFileNode now. I added [better-queue](https://www.npmjs.com/package/better-queue) to core recently and it's fantastic.

**Comment by tsimons** [comment]
I will look in to that for tomorrow. I definitely want to be consistent with existing approaches

**Comment by pieh** [comment]
@tsimons hey, did you have chance to look on doing this in `createRemoteFileNode`? If not I will do it today or over the weekend - just let me know.

**Comment by tsimons** [comment]
@pieh That's actually a much better idea. I was queueing the call to createRemoteFileNode from the WP plugin. I had a bunch of stuff to check before being able to update this. It takes about 15 minutes per build so the feedback loop is insanely long for me.

We run our blog on heroku and the staging server has 1x dyno (500MB RAM really weak compute). I kept getting intermittent 503 if I asked for too much at once, even with queueing.

I also was facing issues with the js heap running out of memory. Running develop would hold about 5GB of RAM, 3GB related to the CSS task, 2GB to the JS task. Another optimization that I think can be made is to utilize disk more than RAM. I haven't gotten a chance to look under the hood of gatsby core yet, but it would be a nice to stream the creation of nodes and free up memory when they are done.

**Comment by tsimons** [comment]
I also had to update to node v9.8 because the sourcemaps were creating a string larger than 256MB and V8 didn't allow strings of that size. Last year they bumped it up to 1GB: https://bugs.chromium.org/p/v8/issues/detail?id=6148#c10

**Comment by tsimons** [comment]
Worked on adding the queue to createRemoteFileNode. I didn't get much of a chance to fully test it though. The unit tests pass, but trying to actually run it takes too long for me and I have to head out.

I wanted to push something up so you wouldn't have to start from scratch if you end up working on this over the weekend.

**Comment by pieh** [comment]
hey @tsimons I didn't manage to work on it during the weekend, but just few notes for what You did already:
- this should be split for 2 separate PRs - one for `createRemoteFileNode` and one for wordpress plugin
- re feedback loop - for developing it's best to create minimal wordpress instance and test it against your real site once you get it to work with minimal one
- using `batchSize` vs `concurent` in `better-queue` - I may be wrong here but doesn't batchSize "just" take tasks and chunks them into arrays and then you need to handle those array in your queue handler function?

**Comment by KyleAMathews** [comment]
Yeah, `concurrent` is what we want — batching isn't what we want since each download is done separately.

**Comment by tsimons** [comment]
Opened a separate PR for the WP fetch batching: https://github.com/gatsbyjs/gatsby/pull/4735

**Comment by tsimons** [comment]
OK, looks like this now working. I have the queue setup for create remote filenode in the source FS module. There is a tradeoff here however. The max concurrency is statically set to 200. How would you go about raising that? `ulimit -n` isn't a good option since it doesn't work well on a lot of systems, be it Windows or CI systems.

What is your recommended approach to make that value confgurable. Keep in mind that the queue is instantiated before the call to createRemoteFileNode

**Comment by KyleAMathews** [comment]
Not sure — but I'd assume that 200 should be high enough? That should easily saturate the network for pretty much everyone.

**Comment by KyleAMathews** [comment]
It is tricky configuring shared helper functions like this. If we really do need to add configurability, we could perhaps use an ENV variable (e.g. for use on a CI server that has a super fast network). But I really think we don't need to so let's just ignore the issue for now.

**Comment by tsimons** [comment]
Sounds great! This is good to go on my end. Merge when you feel it's ready. I did change quite a bit in the file to reduce complexity and add inline docs, so I understand if it takes a little while

**Comment by tsimons** [comment]
Any update on this? I have to use `gatsby-dev` until this is merged

File: packages/gatsby-source-filesystem/src/create-remote-file-node.js [hunk 1]
```diff
@@ -3,112 +3,283 @@ const got = require(`got`)
...
+
+/**
+ * @typedef {Auth}
+ * @type {Object}
+ * @property {String} htaccess_pass
+ * @property {String} htaccess_user
+ */
+
+/**
+ * @typedef {CRFNPayload}
```
Comment (pieh) [inline packages/gatsby-source-filesystem/src/create-remote-file-node.js]: Any reason you shorten type prefix to CRFN? Not a blocker for me, just curious if we couldn't use full name here so anyone using it later don't have to decipher abbreviation.

File: packages/gatsby-source-filesystem/src/create-remote-file-node.js [hunk 2]
```diff
@@ -3,112 +3,283 @@ const got = require(`got`)
...
+ */
+module.exports = ({ url, store, cache, createNode, auth = {} }) => {
+  // Check if we already requested node for this remote file
+  // and return stored promise if we did.
+  if (processingCache[url]) {
+    return processingCache[url]
+  }
+
+
+
```
Comment (pieh) [inline packages/gatsby-source-filesystem/src/create-remote-file-node.js]: just cosmetic - can you remove multi empty lines?

File: packages/gatsby-source-filesystem/src/create-remote-file-node.js (hunk 1 above)
Comment (tsimons) [reply packages/gatsby-source-filesystem/src/create-remote-file-node.js]: It just felt super long, particularly when using it in a param tag. I'll update it

File: packages/gatsby-source-filesystem/src/create-remote-file-node.js (hunk 2 above)
Comment (tsimons) [reply packages/gatsby-source-filesystem/src/create-remote-file-node.js]: Sure thing

**Comment by tsimons** [comment]
@pieh Changes have been pushed

**Comment by KyleAMathews** [comment]
Thanks! This is a great PR

**Comment by jlengstorf** [comment]
Hiya @tsimons! 👋

This is definitely late, but on behalf of the entire Gatsby community, I wanted to say thank you for being here.

Gatsby is built by awesome people like you. Let us say “thanks” in two ways:

1.  **We’d like to send you some Gatsby swag.** As a token of our appreciation, you can go to the [Gatsby Swag Store][store] and log in with your GitHub account to get a coupon code good for one free piece of swag. (We’ve got t-shirts and hats, plus some socks that are really razzing our berries right now.)
2.  **If you’re not already part of it, we just invited you to join the Gatsby organization on GitHub.** This will add you to our team of maintainers. You’ll receive an email shortly asking you to confirm. By joining the team, you’ll be able to label issues, review pull requests, and merge approved pull requests.

If you have questions, please don’t hesitate to reach out to us: tweet at [@gatsbyjs][twitter] and we’ll come a-runnin’.

Thanks again! 💪💜

[store]: https://store.gatsbyjs.org
[twitter]: https://twitter.com/gatsbyjs

**Pull request merged**